SLING-11906 Migrate to slf4j 2.x

enapps-enorman opened the pull request:
Increasingly more libraries have been migrating to slf4j 2.x (for example logback 1.3+, tika 2.5+ and jetty 10+)
To be compatible with those, the sling commons log bundle should migrate to slf4j v2.x and logback v1.5.x

The changes proposed here is an attempt to do the refactoring to migrate to the newer versions.

Here are some differences in the proposal:
-----

1. Per: https://www.slf4j.org/faq.html#changesInVersion200

	SLF4J 2.x has switched to use the java.util.ServiceLoader to locate the logging implementation so we now require a service loader mediator implementation in order for the slf4j-api to find the logback implementation.

2. I have tried to avoid embedding the slf4j and logback classes inside of the org.apache.sling.commons.log bundle and instead just use the bundles that were already provided by those other organizations.  That approach mostly worked, but I did encounter one classloading (visiblity) problem from our MaskingMessageUtil customizations where the class names were being passed to logback and loaded via reflection.  I was also able to create a temporary workaround by forking and updating the logback PatternLayout (and related) classes to additionally pass a Supplier instead of a classname for that.

	I proposed a change to logback that would make my forked workaround unnecessary which is tracked at: https://github.com/qos-ch/logback/issues/803

3. There still exists a problem with some of the OAK bundles importing the org.slf4j.event with an old version that is not exported in the SLF4j 2.x implementation.  I was able to workaround this by creating a fragment bundle that is attached to slf4j-api and exports the 1.7.36 version of the org.slf4j.event package.  Hopefully OAK can be modified to remove the dependency on org.slf4j.event?


	For your reference, I stashed my workaround fragment bundle project at: https://github.com/enapps-enorman/sling-org-apache-sling-slf4j1x-compatibility

	 I believe the OAK troubles are tracked as: https://issues.apache.org/jira/browse/OAK-10339

	The result of this problem is errors reported by the feature analyzer like this:

	[ERROR] [bundle-packages] org.apache.jackrabbit:oak-lucene:1.60.0: Bundle is importing package org.slf4j.event;version=[1.7,2) with start order 15 but no bundle is exporting these for that start order in the required version range.
	[ERROR] [bundle-packages] org.apache.jackrabbit:oak-store-document:1.60.0: Bundle is importing package org.slf4j.event;version=[1.7,2) with start order 15 but no bundle is exporting these for that start order in the required version range.
	[ERROR] [bundle-packages] org.apache.jackrabbit:oak-commons:1.60.0: Bundle is importing package org.slf4j.event;version=[1.7,2) with start order 15 but no bundle is exporting these for that start order in the required version range.
	[ERROR] [bundle-packages] org.apache.jackrabbit:oak-core:1.60.0: Bundle is importing package org.slf4j.event;version=[1.7,2) with start order 15 but no bundle is exporting these for that start order in the required version range.
	[ERROR] [bundle-packages] org.apache.jackrabbit:oak-query-spi:1.60.0: Bundle is importing package org.slf4j.event;version=[1.7,2) with start order 15 but no bundle is exporting these for that start order in the required version range.

4. Integrating this into the starter (or other sling based application) requires something like the following set of bundles at the top of the boot.json feature descriptor:

	For example: https://github.com/apache/sling-org-apache-sling-starter/compare/master...enapps-enorman:sling-org-apache-sling-starter:issue/SLING-11906

        {
            "id":"org.ow2.asm:asm-analysis:${asm.version}",
            "start-order":"1"
        },
        {
            "id":"org.ow2.asm:asm-commons:${asm.version}",
            "start-order":"1"
        },
        {
            "id":"org.ow2.asm:asm-tree:${asm.version}",
            "start-order":"1"
        },
        {
            "id":"org.ow2.asm:asm-util:${asm.version}",
            "start-order":"1"
        },
        {
            "id":"org.ow2.asm:asm:${asm.version}",
            "start-order":"1"
        },
        {
            "id":"org.apache.aries.spifly:org.apache.aries.spifly.dynamic.bundle:1.3.7",
            "start-order":"1"
        },
        {
            "id":"org.apache.sling:org.apache.sling.commons.log:6.0.0-SNAPSHOT",
            "start-order":"1"
        },
        {
            "id":"ch.qos.logback:logback-core:${logback.version}",
            "start-order":"1"
        },
        {
            "id":"ch.qos.logback:logback-classic:${logback.version}",
            "start-order":"1"
        },
        {
            "id":"org.slf4j:jul-to-slf4j:${slf4j.version}",
            "start-order":"1"
        },
        {
            "id":"org.slf4j:jcl-over-slf4j:${slf4j.version}",
            "start-order":"1"
        },
        {
            "id":"org.slf4j:log4j-over-slf4j:${slf4j.version}",
            "start-order":"1"
        },
        {
            "id":"org.apache.sling:slf4j1x-compatibility:0.9-SNAPSHOT",
            "start-order":"1"
        },
        {
            "id":"org.slf4j:slf4j-api:${slf4j.version}",
            "start-order":"1"
        },

5. In logback 1.3 or later, the ch.qos.logback.classic.jmx package was removed for security reasons and for lack of use.

	Per: https://logback.qos.ch/news.html#1.3.0

	If anyone has a custom logback.xml file, then the `<jmxConfigurator />` element isn't supported by logback anymore and will need to be removed.

6. Also, in logback.xml file, the /configuration/*osgi rules don't need to be declared as <newRule>
elements anymore as these are added automatically now.

        <newRule pattern="*/configuration/osgi"
             actionClass="org.apache.sling.commons.log.logback.OsgiAction"/>
        <newRule pattern="*/configuration/appender-ref-osgi"
             actionClass="org.apache.sling.commons.log.logback.OsgiAppenderRefAction"/>

reschke commented:
For the oak commons issue, I have started work in https://issues.apache.org/jira/browse/OAK-10755 - this deprecates the old API, and adds a new one which does not need org.slf4j.event anymore. This would go into Oak 1.64.0. In the next release, we would then remove the deprecated method signature.

rombert commented:
Thanks a lot for preparing this @enapps-enorman ! I'll try to take a look in the coming days.

rombert commented on src/main/java/ch/qos/logback/classic/PatternLayoutOsgi.java:
```diff
@@ -0,0 +1,183 @@
+import java.util.HashMap;
+import java.util.Map;
+
+import ch.qos.logback.classic.pattern.*;
```
@rishabhdaim - thanks for reviewing this PR. IIUC, @enapps-enorman has 'vendored'/copied over code from a third party library. If we decide to go this way I think we should keep it unchanged, deviations make it hard to compare with upstream and resync with later versions, if needed.

rombert commented on src/main/java/ch/qos/logback/classic/PatternLayoutOsgi.java:
```diff
@@ -0,0 +1,183 @@
+
+public class PatternLayoutOsgi extends PatternLayoutBaseOsgi<ILoggingEvent> {
+
+    public static final Map<String, String> DEFAULT_CONVERTER_MAP = new HashMap<>();
```
Same as https://github.com/apache/sling-org-apache-sling-commons-log/pull/18#discussion_r1574507607

rombert reviewed (commented):
Again, thanks a lot for preparing this PR @enapps-enorman .

Before diving into the details, I want start discussion on two of the high-level topics from your PR

1. ServiceLoder (+mediator)

We have historically stayed away from ServiceLoader usage, because it's a bit complicated (even though it works) in OSGi environments but also because we don't have a real dependency to provide; in case of `jakarta.json`  we provide a single Commons Johnzon implemantation, and in case of `slf4j` we provide our own Commons Logging.

Would it be possible to short-circuit the ServiceLoader in a similar manner to the work we do for Commons Johnzon? That would keep deployments lean and do less unneeded work.

2. The classloading issues from Logback

It would, of course, be prefferable to have upstream fix your improvements so we don't need to fork. In  case that does not happen, are there any alternatives we can use, like reflection and/or bytecode weaving? We did make use of reflection on the Sling XSS bundle to fix certain issues in upstream libraries we were embedding. It is perhaps not the nicest code, but it helped us not fork + vendor code when upstream was not open to those changes.

rombert commented:
> For your reference, I stashed my workaround fragment bundle project at: https://github.com/enapps-enorman/sling-org-apache-sling-slf4j1x-compatibility

On that topic, I think the bundle may be useful in case we are ready with the switch before Oak is, thank you for preparing that.

enapps-enorman commented:
> 1. ServiceLoder (+mediator)
...
> Would it be possible to short-circuit the ServiceLoader in a similar manner to the work we do for Commons Johnzon? That would keep deployments lean and do less unneeded work.

The ServiceLoader mediator availability isn't really a problem for me since I have to have it around anyway for the various jetty-* bundles that I use that require it.  Also, the sling starter already includes the mediator bundles for groovy, so...

In my opinion, trying to do that kind of ServiceLoader workaround is a losing battle.  I think what you are suggesting is to repackage the ServiceLoader consumers and providers into the same bundle so there is no classloading visibility problems.  The problem with that is that someone has to be paying attention whenever any of those third-party bundles has a new release and then do a new release of the o.a.sling.commons.log with the new bits.  That didn't happen the last time which led to the embedded slf4j / logback classes in the o.a.sling.commons.log bundle drifting years out of date.

With that said, I do see that there is one other option described in the slf4j FAQ that might work without the mediator: https://www.slf4j.org/faq.html#changesInVersion200

>  SINCE 2.0.9 You can specify the provider class explicitly via the "slf4j.provider" system property. This bypasses the service loader mechanism for finding providers and may shorten SLF4J initialization.

I believe the extra work to do that would be to configure the "slf4j.provider" framework property, export the "org.apache.sling.commons.log.logback.spi" package from the o.a.sling.commons.log bundle and then deploy a fragment bundle attached to the slf4j-api host bundle that augments the Import-Package instruction to include the "org.apache.sling.commons.log.logback.spi" package to make it visible to the slf4j-api classloader.

If that works then the choice to use the ServiceLoader mediator or the framework-property + fragment bundle could be decided by whoever is constructing the application.  Let me know if that would be acceptable to you.

> 2. The classloading issues from Logback

It has been 2 weeks since I opened the PR proposal to logback and there has been no response at all.  Do you know anyone over there would could review it and give a yes or no answer?

I'm not really interested in wasting more of my time on alternate ways to do the same thing.  If you want to try some other approach then please provide a patch with the changes you think will work better.

enapps-enorman commented:
> slf4j.provider

@rombert  After further review, using slf4j-api without the service mediator as described in my last comment would also require changes to the slf4j-api manifest as it currently does not declare the serviceloader capabilities as optional.

For example:
`Require-Capability: 
osgi.extender;filter:="(&(osgi.extender=osgi.serviceloader.processor)(version>=1.0.0)(!(version>=2.0.0)))",
osgi.serviceloader;filter:="(osgi.serviceloader=org.slf4j.spi.SLF4JServiceProvider)";osgi.serviceloader="org.slf4j.spi.SLF4JServiceProvider"`

rombert commented:
@enapps-enorman

Regarding the slf4j-api - could we not also register the capability ourselves, since we provide it and bypass the serviceloader?

Regarding the logback contribution - I am not sure if/how we can expedite the logback contribution. At this point I am wary of importing a large number of source files in the Sling Git repository; vendoring is not that easy and you can see on https://github.com/apache/sling-org-apache-sling-commons-json/tree/maintenance how we need to make changes and track them. 

I fully understand your desire to not do more work unless really required. I can think of alternative ways of approaching the logback issue, in case someone has the desire to explore them. We ended up using reflection in the Sling XSS bundle when upstream was unresponsive ( https://github.com/OWASP/java-html-sanitizer/issues/275 ) but we could also consider using Weaving Hooks ( https://docs.osgi.org/specification/osgi.core/7.0.0/framework.weavinghooks.html ) .

enapps-enorman commented:
> Regarding the slf4j-api - could we not also register the capability ourselves, since we provide it and bypass the serviceloader?

It seems like that would interfere with other usages of the serviceloader so I am not inclined to spend any time on that.

> At this point I am wary of importing a large number of source files in the Sling Git repository

I'm not sure what you mean by a "large number" of source files?  There are only 4 files under src/main/java/ch/qos/logback and only 1 of those files copies any code blocks from the original vendor.  The other three extend the original classes and add to it or override to provide a different implementation.

Anyways with that said, I am not interested in spending any more time on solving this in another way.  If you want to do that work then please go ahead and make a proposal to prove your theory.

I'm perfectly happy just managing and using my own fork at this point since I don't anticipate it needing much maintenance in the future.

enapps-enorman commented:
FYI: A variation of the changes I requested in https://github.com/qos-ch/logback/issues/803 has been applied to the logback codebase.  This means that whenever the next public release of logback happens (ETA unknown) then we can refactor this PR to remove the 4 files under src/main/java/ch/qos/logback

enapps-enorman commented:
FYI: logback 1.5.13 has now been released which contains the changes I requested in https://github.com/qos-ch/logback/issues/803

I have updated this PR to depend on that new version of logback and removed the workaround ch/qos/logback/* classes.  I also refactored any references to those workaround classes to go back to the original equivalents.

I hope this has addressed some of the concerns.

At this point I would request the interested parties to review the changes again and provide fresh feedback if there is any remaining reasons to not merge the changes.

rombert commented:
Thank you for your persistence @enapps-enorman . This is on my to-do list but given the size of the chage ( 14k lines added, 4 removed ) it is going to take some time. I don't expect to be done with this in December. Of course, that should not stop anyone else from reviewing, I don't have to be the reviewer :-)

Appreciate the effort that you put in the PR and looking forward to finally having slf4j 2.0 in Sling.

rombert commented:
@enapps-enorman - the changes look very good to me. Happy to see that we're not embedding any bundles anymore, which will make upgrades much easier and the need to release new versions of this bundle much smaller.

I think the version bump of the exported `org.apache.sling.commons.log.logback` is fine.

I am still thinking of the SPIFly requirement and would like to avoid it. I think for scenarions like the sling-mini it would be very good to keep the bundle requirements low. I am not going to oppose merging as-is but will write an email to dev@sling to give everyone the change to comment or - ideally - to come up with a way of avoiding adding this requirement.

rombert commented:
> I am still thinking of the SPIFly requirement and would like to avoid it. I think for scenarions like the [sling-mini](https://issues.apache.org/jira/browse/SLING-mini) it would be very good to keep the bundle requirements low. I am not going to oppose merging as-is but will write an email to dev@sling to give everyone the change to comment or - ideally - to come up with a way of avoiding adding this requirement.

https://lists.apache.org/thread/ro2bno850w8z7g8rtjh832h05698n8xn

enapps-enorman commented:
FYI: It looks to me that the latest version (1.76.0 or maybe earlier) of OAK has fixed the "importing the org.slf4j.event with an old version" problem so the slf4j1x-compatibility workaround fragment bundle isn't needed in the "starter" distribution anymore.

```
 {
     "id":"org.apache.sling:slf4j1x-compatibility:0.9-SNAPSHOT",
     "start-order":"1"
 },
```

cziegeler commented:
How about we merge this PR as-is and then figure out how to handle the spifly dependency in another change before the release? I think we should value @enapps-enorman great work (thanks!) and move forward.

rombert commented:
Agreed @cziegeler . I resolved the merge conflicts, will merge the PR as soon as CI passes.

rombert commented:
Again, thanks a lot @enapps-enorman for the large effort in making this work.
